### PR TITLE
Add Earthly configuration and use it to build the spaceros base image.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
This PR has been updated so Earthly completely replaces the base docker image.

This work spanned several other significant transitions for Space ROS, including a transition from ROS 2 Rolling as the foundation to ROS 2 Humble, and the more impactful change from Ubuntu 20.04 to Ubuntu 22.04 as the base platform.

The impacts of the latter change present problems [building ikos](https://github.com/NASA-SW-VnV/ikos/issues/193) and with availability of the [nvidia/cudagl image](https://gitlab.com/nvidia/container-images/cuda/-/issues/147) in limbo right now it's unlikely there will be official 22.04-based images in the near future.

For those reasons the current image does not build or use ikos or bundle nvidia drivers.

I was always planning to use a standard Ubuntu base image for the base space-ros build and take advantage of multi-stage image builds to copy artifacts from that base image to a new demo image based on the nvidia one, however since the nvidia images aren't available for 22.04 using the [`rocker`](https://github.com/osrf/rocker) utility to inject nvidia support at the time its needed is a worthy alternative and I've started updating the README.md in that direction. Since the base Space ROS image doesn't have any included GUI utilities there's not much to add for this README.

Not yet in the README.md is general information on how to use earthly itself rather than just as the docker image build interface. The documentation for that is included as part of https://github.com/space-ros/docs/pull/5 and if anything this README may just link to that document once it is published.

### CODE REVIEW

**Comment by mjeronimo** [comment]
Worked great after that fix. I like the concept of Earthly being a combination of makefiles/Docker; it makes it pretty intuitive for me (no pun intended).  One thing that wasn't entirely clear to me was when the docker image gets created. I assume that everything lives in the cache until then?

**Comment by nuclearsandwich** [comment]
> One thing that wasn't entirely clear to me was when the docker image gets created. I assume that everything lives in the cache until then?

Earthly uses a moby project called buildkitd, which is an improved container image builder for docker (and accessible in recent docker releases as `docker buildx`. The Earthly buildkit daemon does not use your system's docker image cache or container framework (that is, you won't see earthly containers in `docker ps` or images in `docker images`) but it does maintain its own image cache as a docker volume attached to the buildkitd container that it will spawn on startup.

**Comment by nuclearsandwich** [comment]
This PR is still in draft status and one of the reasons is trying to determine whether we should start to use Earthly now or wait until the second demo with Dagger so that we don't migrate the current image build twice. I'm not sure that Dagger's tooling offers the same kind of fundamental replacement for the Dockerfile-based image builds that Earthly does (and the image building was not the Earthly feature that we were actually seeking to evaluate).

I see a few possibilities.
* We could make the Earthly configuration and project its own repository and archive it if we don't move forward with it.
* We could switch all our Docker image builds to Earthly knowing that it's a nice tool for the job whether or not we continue to use it for other parts of the CI/CD infra for Space ROS
* We could leave this PR in draft pending a Dagger demo sometime after 15 August.

**Comment by nuclearsandwich** [comment]
While it is possible to reference Earthly targets in other repositories, I'm not sure how well that fits in our workflow or where the "correct" place for Earthly build files for common reference should be. 

I haven't come to any conclusions yet but I have opened a similar PR in the demos repo https://github.com/space-ros/demos/pull/11 which contains some common functionality.

File: spaceros/Earthfile [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    python3-pip \
+    python3-pytest \
+    python3-pytest-cov \
+    python3-pytest-repeat \
+    python3-pytest-rerunfailures \
+    python3-rosdep \
+    python3-setuptools \
+    python3-vcstool \
+    wget \
+    vim
```
Comment (nuclearsandwich) [inline spaceros/Earthfile:70]: We can replace a lot of this with ros-dev-tools and drop much of the rest once https://github.com/ament/ament_lint/pull/420 is released.

File: spaceros/README.md [hunk 2]
```diff
@@ -78,12 +80,21 @@ Commands:
...
+```
+spaceros-user@d10d85c68f0e:~/spaceros$ mkdir src
+spaceros-user@d10d85c68f0e:~/spaceros$ vcs import src < exact.repos
+```
 
+From there you can run a new build and any additional tests.
 ```
-spaceros-user@d10d85c68f0e:~/spaceros$ colcon test
+spaceros-user@d10d85c68f0e:~/spaceros$ colcon build --cmake-args -DCMAKE_BUILD_TYPE=Debug --DCMAKE_EXPORT_COMPILE_COMMANDS=ON --no-warn-unused-cli
+spaceros-user@d10d85c68f0e:~/spaceros$ colcon test --ctest-args -LE "(ikos|xfail)" --pytest-args -m "not xfail"
```
Comment (nuclearsandwich) [inline spaceros/README.md:97]: @mjeronimo this whole section may be a bit of an overstep for this README. I think moving it to the docs site either in https://github.com/space-ros/docs/pull/5 or another tutorial specifically about the docker image may make sense.

If we do want to leave it here, I updated the `colcon test` command to match the args used during automated builds (minus retests).

Comment (mjeronimo) [reply spaceros/README.md:97]: We can work out the proper split when we update the documentation site. I've been thinking about this topic and what information goes where. Good to have it here for now and we can refactor.

**Review by mjeronimo — APPROVED** [review]

**Pull request merged**